diff --git a/config/BUILD.gn b/config/BUILD.gn
index 35c9a6275..37fd23e98 100644
--- a/config/BUILD.gn
+++ b/config/BUILD.gn
@@ -289,7 +289,7 @@ group("shared_library_deps") {
 # Executable configs -----------------------------------------------------------
 
 # Windows linker setup for EXEs and DLLs.
-if (is_win) {
+if (is_win && !use_zig_tc) {
   _windows_linker_configs = [
     "//build/config/win:sdk_link",
     "//build/config/win:common_linker_setup",
@@ -300,9 +300,9 @@ if (is_win) {
 config("executable_config") {
   configs = []
 
-  if (is_win) {
+  if (is_win && !use_zig_tc) {
     configs += _windows_linker_configs
-  } else if (is_mac) {
+  } else if (is_mac && !use_zig_tc) {
     configs += [ "//build/config/mac:mac_dynamic_flags" ]
   } else if (is_ios) {
     configs += [
@@ -332,9 +332,9 @@ config("executable_config") {
 config("shared_library_config") {
   configs = []
 
-  if (is_win) {
+  if (is_win && !use_zig_tc) {
     configs += _windows_linker_configs
-  } else if (is_mac) {
+  } else if (is_mac && !use_zig_tc) {
     configs += [ "//build/config/mac:mac_dynamic_flags" ]
   } else if (is_ios) {
     configs += [
diff --git a/config/BUILDCONFIG.gn b/config/BUILDCONFIG.gn
index cc7e92ddb..2d1db7e2a 100644
--- a/config/BUILDCONFIG.gn
+++ b/config/BUILDCONFIG.gn
@@ -158,6 +158,15 @@ declare_args() {
   # DON'T ADD MORE FLAGS HERE. Read the comment above.
 }
 
+declare_args() {
+  use_zig_tc = false
+  cxx_use_ld = "ldd"
+  zig_cc = ""
+  zig_cxx = ""
+  host_zig_cc = ""
+  host_zig_cxx = ""
+}
+
 declare_args() {
   # Debug build. Enabling official builds automatically sets is_debug to false.
   is_debug = !is_official_build
@@ -354,7 +363,7 @@ default_compiler_configs = [
   "//build/config/sanitizers:default_sanitizer_flags",
 ]
 
-if (is_win) {
+if (is_win && !use_zig_tc) {
   default_compiler_configs += [
     "//build/config/win:default_cfg_compiler",
     "//build/config/win:default_crt",
@@ -411,7 +420,7 @@ set_defaults("rust_proc_macro") {
 
 # Compute the set of configs common to all linked targets (shared libraries,
 # loadable modules, executables) to avoid duplication below.
-if (is_win) {
+if (is_win && !use_zig_tc) {
   # Many targets remove these configs, so they are not contained within
   # //build/config:executable_config for easy removal.
   _linker_configs = [
@@ -421,7 +430,7 @@ if (is_win) {
     # that shouldn't use the windows subsystem.
     "//build/config/win:console",
   ]
-} else if (is_mac) {
+} else if (is_mac && !use_zig_tc) {
   _linker_configs = [ "//build/config/mac:strip_all" ]
 } else {
   _linker_configs = []
@@ -433,7 +442,7 @@ default_executable_configs = default_compiler_configs + [
                                "//build/config:executable_config",
                              ] + _linker_configs
 
-if (is_win) {
+if (is_win && !use_zig_tc) {
   # Turn on linker CFI for executables, and position it so it can be removed
   # if needed.
   default_executable_configs += [ "//build/config/win:cfi_linker" ]
@@ -449,7 +458,7 @@ default_shared_library_configs = default_compiler_configs + [
                                    "//build/config:default_libs",
                                    "//build/config:shared_library_config",
                                  ] + _linker_configs
-if (is_win) {
+if (is_win && !use_zig_tc) {
   # Turn on linker CFI for DLLs, and position it so it can be removed if needed.
   default_shared_library_configs += [ "//build/config/win:cfi_linker" ]
 }
diff --git a/config/compiler/BUILD.gn b/config/compiler/BUILD.gn
index 50be57ff8..ddac52169 100644
--- a/config/compiler/BUILD.gn
+++ b/config/compiler/BUILD.gn
@@ -263,7 +263,7 @@ config("compiler") {
   # System-specific flags. If your compiler flags apply to one of the
   # categories here, add it to the associated file to keep this shared config
   # smaller.
-  if (is_win) {
+  if (is_win && !use_zig_tc) {
     configs += [ "//build/config/win:compiler" ]
   } else if (is_android) {
     configs += [ "//build/config/android:compiler" ]
@@ -271,7 +271,7 @@ config("compiler") {
     configs += [ "//build/config/linux:compiler" ]
   } else if (is_nacl) {
     configs += [ "//build/config/nacl:compiler" ]
-  } else if (is_mac) {
+  } else if (is_mac && !use_zig_tc) {
     configs += [ "//build/config/mac:compiler" ]
   } else if (is_ios) {
     configs += [ "//build/config/ios:compiler" ]
@@ -292,6 +292,11 @@ config("compiler") {
     ":compiler_deterministic",
   ]
 
+  if (use_zig_tc) {
+    # We'll provide the target flags from build.zig.
+    configs -= [ ":compiler_cpu_abi" ]
+  }
+
   # Here we enable -fno-delete-null-pointer-checks, which makes various nullptr
   # operations (e.g. dereferencing) into defined behavior. This avoids deletion
   # of some security-critical code: see https://crbug.com/1139129.
@@ -344,7 +349,7 @@ config("compiler") {
     }
 
     if (use_lld) {
-      ldflags += [ "-fuse-ld=lld" ]
+      ldflags += [ "-fuse-ld=${cxx_use_ld}" ]
       if (lld_path != "") {
         ldflags += [ "-B$lld_path" ]
       }
@@ -562,7 +567,7 @@ config("compiler") {
     }
 
     # TODO(crbug.com/1235145): Investigate why/if this should be needed.
-    if (is_win) {
+    if (is_win && !use_zig_tc) {
       cflags += [ "/clang:-ffp-contract=off" ]
     } else {
       cflags += [ "-ffp-contract=off" ]
@@ -621,7 +626,7 @@ config("compiler") {
       # support "c++20"/"gnu++20".
       cflags_cc += [ "-std=${standard_prefix}++2a" ]
     }
-  } else if (is_win) {
+  } else if (is_win && !use_zig_tc) {
     cflags_c += [ "/std:c11" ]
     if (use_cxx17 || (!is_clang && defined(msvc_use_cxx17) && msvc_use_cxx17)) {
       cflags_cc += [ "/std:c++17" ]
@@ -1473,7 +1478,7 @@ config("runtime_library") {
   # System-specific flags. If your compiler flags apply to one of the
   # categories here, add it to the associated file to keep this shared config
   # smaller.
-  if (is_win) {
+  if (is_win && !use_zig_tc) {
     configs += [ "//build/config/win:runtime_library" ]
   } else if (is_linux || is_chromeos) {
     configs += [ "//build/config/linux:runtime_library" ]
@@ -1483,7 +1488,9 @@ config("runtime_library") {
   } else if (is_ios) {
     configs += [ "//build/config/ios:runtime_library" ]
   } else if (is_mac) {
-    configs += [ "//build/config/mac:runtime_library" ]
+    if (!use_zig_tc) {
+      configs += [ "//build/config/mac:runtime_library" ]
+    }
   } else if (is_android) {
     configs += [ "//build/config/android:runtime_library" ]
   }
@@ -1504,7 +1511,7 @@ config("default_warnings") {
   cflags_cc = []
   ldflags = []
 
-  if (is_win) {
+  if (is_win && !use_zig_tc) {
     if (treat_warnings_as_errors) {
       cflags += [ "/WX" ]
     }
@@ -1684,7 +1691,7 @@ config("prevent_unsafe_narrowing") {
 # part of Chromium.
 
 config("chromium_code") {
-  if (is_win) {
+  if (is_win && !use_zig_tc) {
     if (is_clang) {
       cflags = [ "/W4" ]  # Warning level 4.
 
@@ -1771,7 +1778,7 @@ config("no_chromium_code") {
   cflags_cc = []
   defines = []
 
-  if (is_win) {
+  if (is_win && !use_zig_tc) {
     if (is_clang) {
       cflags += [ "/W3" ]  # Warning level 3.
     }
@@ -1848,7 +1855,7 @@ config("rtti") {
 config("no_rtti") {
   # Some sanitizer configs may require RTTI to be left enabled globally
   if (!use_rtti) {
-    if (is_win) {
+    if (is_win && !use_zig_tc) {
       cflags_cc = [ "/GR-" ]
     } else {
       cflags_cc = [ "-fno-rtti" ]
@@ -1890,7 +1897,7 @@ config("thin_archive") {
   # archives.
   # TODO(crbug.com/1221615): Enable on is_apple if use_lld once that no longer
   # confuses lldb.
-  if ((is_posix && !is_nacl && !is_apple) || is_fuchsia) {
+  if ((is_posix && !is_nacl && !is_apple) || is_fuchsia || use_zig_tc) {
     arflags = [ "-T" ]
   } else if (is_win && use_lld) {
     arflags = [ "/llvmlibthin" ]
@@ -1982,7 +1989,7 @@ config("no_incompatible_pointer_warnings") {
 
 # Shared settings for both "optimize" and "optimize_max" configs.
 # IMPORTANT: On Windows "/O1" and "/O2" must go before the common flags.
-if (is_win) {
+if (is_win && !use_zig_tc) {
   common_optimize_on_cflags = [
     "/Ob2",  # Both explicit and auto inlining.
     "/Oy-",  # Disable omitting frame pointers, must be after /O2.
@@ -2088,7 +2095,7 @@ config("default_stack_frames") {
 
 # Default "optimization on" config.
 config("optimize") {
-  if (is_win) {
+  if (is_win && !use_zig_tc) {
     if (chrome_pgo_phase != 2) {
       # Favor size over speed, /O1 must be before the common flags.
       # /O1 implies /Os and /GF.
@@ -2199,7 +2206,7 @@ config("optimize_max") {
     configs = [ "//build/config/nacl:irt_optimize" ]
   } else {
     ldflags = common_optimize_on_ldflags
-    if (is_win) {
+    if (is_win && !use_zig_tc) {
       # Favor speed over size, /O2 must be before the common flags.
       # /O2 implies /Ot, /Oi, and /GF.
       cflags = [ "/O2" ] + common_optimize_on_cflags
@@ -2232,7 +2239,7 @@ config("optimize_speed") {
     configs = [ "//build/config/nacl:irt_optimize" ]
   } else {
     ldflags = common_optimize_on_ldflags
-    if (is_win) {
+    if (is_win && !use_zig_tc) {
       # Favor speed over size, /O2 must be before the common flags.
       # /O2 implies /Ot, /Oi, and /GF.
       cflags = [ "/O2" ] + common_optimize_on_cflags
diff --git a/config/linux/BUILD.gn b/config/linux/BUILD.gn
index 131bb71d1..5cde9fff4 100644
--- a/config/linux/BUILD.gn
+++ b/config/linux/BUILD.gn
@@ -42,7 +42,10 @@ config("runtime_library") {
 
   if ((!is_chromeos || default_toolchain != "//build/toolchain/cros:target") &&
       (!use_custom_libcxx || current_cpu == "mipsel")) {
-    libs = [ "atomic" ]
+    if (!use_zig_tc) {
+      # -latomic won't be found when using zig tc, it's already compiled from zig libcxx source
+      libs = [ "atomic" ]
+    }
   }
 }
 
diff --git a/config/sysroot.gni b/config/sysroot.gni
index dea380727..2f493a04f 100644
--- a/config/sysroot.gni
+++ b/config/sysroot.gni
@@ -65,7 +65,7 @@ if (sysroot == "") {
                       "string") == "True",
           "Missing sysroot ($sysroot). To fix, run: build/linux/sysroot_scripts/install-sysroot.py --arch=$_script_arch")
     }
-  } else if (is_mac) {
+  } else if (is_mac && !use_zig_tc) {
     import("//build/config/mac/mac_sdk.gni")
     sysroot = mac_sdk_path
   } else if (is_ios) {
diff --git a/toolchain/gcc_toolchain.gni b/toolchain/gcc_toolchain.gni
index 48f931764..21db4578c 100644
--- a/toolchain/gcc_toolchain.gni
+++ b/toolchain/gcc_toolchain.gni
@@ -361,11 +361,13 @@ template("gcc_toolchain") {
       extra_ldflags = ""
     }
 
+    if (!use_zig_tc) {
     if (system_headers_in_deps) {
       md = "-MD"
     } else {
       md = "-MMD"
     }
+    }
 
     enable_linker_map = defined(invoker.enable_linker_map) &&
                         invoker.enable_linker_map && generate_linker_map
@@ -380,7 +382,11 @@ template("gcc_toolchain") {
     tool("cc") {
       depfile = "{{output}}.d"
       precompiled_header_type = "gcc"
-      command = "$cc $md -MF $depfile ${rebuild_string}{{defines}} {{include_dirs}} {{cflags}} {{cflags_c}}${extra_cppflags}${extra_cflags} -c {{source}} -o {{output}}"
+      if (use_zig_tc) {
+        command = "$cc ${rebuild_string}{{defines}} {{include_dirs}} {{cflags}} {{cflags_c}}${extra_cppflags}${extra_cflags} -c {{source}} -o {{output}}"
+      } else {
+        command = "$cc $md -MF $depfile ${rebuild_string}{{defines}} {{include_dirs}} {{cflags}} {{cflags_c}}${extra_cppflags}${extra_cflags} -c {{source}} -o {{output}}"
+      }
       depsformat = "gcc"
       description = "CC {{output}}"
       outputs = [ "$object_subdir/{{source_name_part}}.o" ]
@@ -389,7 +395,11 @@ template("gcc_toolchain") {
     tool("cxx") {
       depfile = "{{output}}.d"
       precompiled_header_type = "gcc"
-      command = "$cxx $md -MF $depfile ${rebuild_string}{{defines}} {{include_dirs}} {{cflags}} {{cflags_cc}}${extra_cppflags}${extra_cxxflags} -c {{source}} -o {{output}}"
+      if (use_zig_tc) {
+        command = "$cxx ${rebuild_string}{{defines}} {{include_dirs}} {{cflags}} {{cflags_cc}}${extra_cppflags}${extra_cxxflags} -c {{source}} -o {{output}}"
+      } else {
+        command = "$cxx $md -MF $depfile ${rebuild_string}{{defines}} {{include_dirs}} {{cflags}} {{cflags_cc}}${extra_cppflags}${extra_cxxflags} -c {{source}} -o {{output}}"
+      }
       depsformat = "gcc"
       description = "CXX {{output}}"
       outputs = [ "$object_subdir/{{source_name_part}}.o" ]
@@ -398,7 +408,11 @@ template("gcc_toolchain") {
     tool("asm") {
       # For GCC we can just use the C compiler to compile assembly.
       depfile = "{{output}}.d"
-      command = "$asm $md -MF $depfile ${rebuild_string}{{defines}} {{include_dirs}} {{asmflags}}${extra_asmflags} -c {{source}} -o {{output}}"
+      if (use_zig_tc) {
+        command = "$asm ${rebuild_string}{{defines}} {{include_dirs}} {{asmflags}}${extra_asmflags} -c {{source}} -o {{output}}"
+      } else {
+        command = "$asm $md -MF $depfile ${rebuild_string}{{defines}} {{include_dirs}} {{asmflags}}${extra_asmflags} -c {{source}} -o {{output}}"
+      }
       depsformat = "gcc"
       description = "ASM {{output}}"
       outputs = [ "$object_subdir/{{source_name_part}}.o" ]
@@ -412,7 +426,11 @@ template("gcc_toolchain") {
       } else {
         rspfile = "{{output}}.rsp"
         rspfile_content = "{{inputs}}"
-        command = "\"$ar\" {{arflags}} -r -c -s -D {{output}} @\"$rspfile\""
+        if (use_zig_tc) {
+          command = "$ar {{arflags}} -r -c -s -D {{output}} @\"$rspfile\""
+        } else {
+          command = "\"$ar\" {{arflags}} -r -c -s -D {{output}} @\"$rspfile\""
+        }
       }
 
       # Remove the output file first so that ar doesn't try to modify the
